Multisig/Logic/BluetoothController.swift

```diff
@@ -0,0 +1,152 @@
+//
+//  BluetoothController.swift
+//  Multisig
+//
+//  Created by Moaaz on 7/26/21.
+//  Copyright © 2021 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+import CoreBluetooth
+
+protocol LedgerNanoXControllerDelegate {
+    func bluetoothControllerDidReceive(response: String)
+    func bluetoothControllerDidFailToConnectBluetooth()
+    func bluetoothControllerDidDiscoverDevice()
+    func bluetoothControllerDidDisconnectDevice()
+}
+
+class BluetoothController: NSObject {
+    static let shared = BluetoothController()
+    private var centralManager: CBCentralManager!
+    var delegate: LedgerNanoXControllerDelegate?
+
+    private var devices: [BluetoothDevice] = []
+    private var supportedDevices: [DeviceConstant] = []
+
+    private var supportedDeviceUuids: [CBUUID] { supportedDevices.compactMap { $0.uuid } }
+    private var supportedDeviceNotifyUuids: [CBUUID] { supportedDevices.compactMap { $0.notifyUuid } }
+
+    override init() {
+        super.init()
+    }
+
+    func scan(devices: [DeviceConstant]) {
+        supportedDevices = devices
+        centralManager = CBCentralManager(delegate: self, queue: nil)
+    }
+
+    func stopScan() {
+        centralManager.stopScan()
+    }
+}
+
+extension BluetoothController: CBCentralManagerDelegate {
+    func centralManagerDidUpdateState(_ central: CBCentralManager) {
+        switch central.state {
+        case .poweredOn:
+            centralManager.scanForPeripherals(withServices: supportedDeviceUuids)
+        default:
+            // Bluetooth might be off or not permitted to use
+            delegate?.bluetoothControllerDidFailToConnectBluetooth()
+            break
+        }
+    }
+
+    func centralManager(_ central: CBCentralManager,
+                        didDiscover peripheral: CBPeripheral,
+                        advertisementData: [String : Any],
+                        rssi RSSI: NSNumber) {
+        if deviceFor(peripheral: peripheral) == nil {
+            devices.append(BluetoothDevice(peripheral: peripheral))
+            delegate?.bluetoothControllerDidDiscoverDevice()
+        }
+
+        centralManager.stopScan()
+    }
+
+    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
+        peripheral.delegate = self
+        peripheral.discoverServices(supportedDeviceUuids)
+    }
+
+    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
+        devices.removeAll { p in p.peripheral == peripheral }
+        delegate?.bluetoothControllerDidDisconnectDevice()
+    }
+}
+
+extension BluetoothController: CBPeripheralDelegate {
+    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
```

Comment:
should we handle errors in all the methods?

Reply:
How about create new GSError type and propagate it ?

Reply:
We will add proper errors if needed in the future. I am not sure what kind of error we need to present to users at the moment.